tests/jballerina-unit-test/src/test/resources/test-src/types/tuples/tuple_access_expr.bal

```diff
@@ -277,3 +277,341 @@ function testTupleAccessUsingUnionWithFiniteTypesNegative() {
+
+    foreach var i in 0 ..< intUnionStringTuple.length() {
+        assertEquals(intUnionStringTuple[i], expectedIntTuple[i]);
+        assertEquals(stringUnionBooleanTuple[i], expectedStringTuple[i]);
+        assertEquals(booleanUnionIntTuple[i], expectedBooleanTuple[i]);
+    }
+}
+
+type IntReadonlyTuple ([int...] & readonly)|[string...];
+type StringReadonlyTuple [boolean...]|([string...] & readonly);
+type BooleanReadonlyTuple ([boolean...] & readonly)|[int...];
+type CustomReadonlyTuple [int...]|[string...]|([int, string, boolean] & readonly)|[boolean...];
+
+public function testTupleAccessWithCustomReadonlyUnionTypes() {
+    IntReadonlyTuple intReadonlyTuple = [1, 2, 3];
+    StringReadonlyTuple stringReadonlyTuple = ["string 1", "string 2", "string 3"];
+    BooleanReadonlyTuple booleanReadonlyTuple = [true, false, true];
+    CustomReadonlyTuple customReadonlyTuple = [1, "string 1", true];
+
+    [int...] expectedIntTuple = [1, 2, 3];
+    [string...] expectedStringTuple = ["string 1", "string 2", "string 3"];
+    [boolean...] expectedBooleanTuple = [true, false, true];
+
+    assertEquals(intReadonlyTuple[index], 1);
+    assertEquals(intReadonlyTuple[index + 1], 2);
+    assertEquals(intReadonlyTuple[index + 2], 3);
+
+    assertEquals(stringReadonlyTuple[index], "string 1");
+    assertEquals(stringReadonlyTuple[index + 1], "string 2");
+    assertEquals(stringReadonlyTuple[index + 2], "string 3");
+
+    assertEquals(booleanReadonlyTuple[index], true);
+    assertEquals(booleanReadonlyTuple[index + 1], false);
+    assertEquals(booleanReadonlyTuple[index + 2], true);
+
+    assertEquals(customReadonlyTuple[index], 1);
+    assertEquals(customReadonlyTuple[index + 1], "string 1");
+    assertEquals(customReadonlyTuple[index + 2], true);
+
+    foreach var i in 0 ..< intReadonlyTuple.length() {
+        assertEquals(intReadonlyTuple[i], expectedIntTuple[i]);
+        assertEquals(stringReadonlyTuple[i], expectedStringTuple[i]);
+        assertEquals(booleanReadonlyTuple[i], expectedBooleanTuple[i]);
+    }
+}
+
+function testTupleWithRestTypesAccess() {
+    [int...] t1 = [1, 2, 3, 5];
+    [int, string, boolean...] t2 = [1, "a", true, true];
+
+    var x1 = t1[0];
+    int x2 = x1;
+    assertEquals(x2, 1);
+
+    var x3 = t2[0];
+    int x4 = x3;
+    assertEquals(x4, 1);
+
+    var x5 = t2[3];
+    boolean x6 = x5;
+    assertEquals(x6, true);
+
+    var x7 = t1[index];
+    int x8 = x7;
+    assertEquals(x8, 1);
+
+    var x9 = t2[index];
+    int|string|boolean x10 = x9;
+    assertEquals(x10, 1);
+
+    var x11 = t2[index + 2];
+    int|string|boolean x12 = x11;
+    assertEquals(x12, true);
+}
+
+type CustomTupleWithRestTypes1 [int...];
+type CustomTupleWithRestTypes2 [int, string, boolean...];
+
+function testCustomTupleWithRestTypesAccess() {
+    CustomTupleWithRestTypes1 t1 = [1, 2, 3, 5];
+    CustomTupleWithRestTypes2 t2 = [1, "a", true, true];
+
+    var x1 = t1[0];
+    int x2 = x1;
```

Comment:
Are there tests like `int x = t1[0]` also? Instead of all going through `var`?

Reply:
Added. 1ef87477be4270f73cec77393ed6e541fc0886eb